README.rst
pyeasyga
========

|PyPI| |Build Status| |Coverage Status| |Downloads|


Introduction
------------
A simple and easy-to-use Genetic Algorithm implementation library in Python.

``pyeasyga`` provides a simple interface to the power of Genetic Algorithms
(GAs). You don't have to have expert GA knowledge in order to use it.

* Homepage: https://github.com/remiomosowon/pyeasyga
* PyPI: https://pypi.python.org/pypi/pyeasyga
* Documentation: http://pyeasyga.readthedocs.org.
* Issues / Feedback: https://github.com/remiomosowon/pyeasyga/issues 
* Free software: BSD license


Installation
~~~~~~~~~~~~

At the command line, simply run::

    $ pip install pyeasyga

Or clone this repository and run ``python setup.py install`` from within the project directory. e.g.::

    $ git clone https://github.com/remiomosowon/pyeasyga.git
    $ cd pyeasyga
    $ python setup.py install

For alternative install methods, see the INSTALL file or the Installation 
section in the documentation.


Examples
~~~~~~~~

See the Usage section in the documentation for examples. The example files can 
be found in the ``examples`` directory.


Note
-----

* Currently under active development


.. |PyPI| image:: https://badge.fury.io/py/pyeasyga.png
   :target: http://badge.fury.io/py/pyeasyga

.. |Build Status| image:: https://travis-ci.org/remiomosowon/pyeasyga.png?branch=develop
   :target: https://travis-ci.org/remiomosowon/pyeasyga

.. |Coverage Status| image:: https://coveralls.io/repos/remiomosowon/pyeasyga/badge.png?branch=develop
   :target: https://coveralls.io/r/remiomosowon/pyeasyga?branch=develop

.. |Downloads| image:: https://pypip.in/d/pyeasyga/badge.png
   :target: https://pypi.python.org/pypi/pyeasyga



HISTORY.rst
.. :changelog:

History
-------

v0.3.0
~~~~~~

**2015-04-07**

* Added Python 3.4 support without breaking Python 2 compatibility (thanks to `yasserglez <https://github.com/yasserglez>`_)

v0.2.5
~~~~~~

**2014-07-09**

* Added an example that solves the `8 Queens Puzzle
  <http://en.wikipedia.org/wiki/Eight_queens_puzzle>`_

**2014-07-09**

* Modified the GeneticAlgorithm class initialisation parameters
* Made changes to USAGE documentation
* Added EXAMPLE documentation as a separate section

v0.2.4
~~~~~~

**2014-07-07**

* Refactored most of the code; Made GeneticAlgorithm class more OOP
* Made changes to INSTALLATION documentation

v0.2.3
~~~~~~

**2014-07-05**

* Fixed breaking python 2.6 build

v0.2.2
~~~~~~

**2014-07-05**

* Removed duplicate ‘Example’ documentation; now maintaining only one copy in 
  examples/README.rst
* Added link to jeffknupp’s sandman repo in HISTORY
* Modified release option in Makefile to also upload project documentation
* Added INSTALLATION and EXAMPLE sections to README.rst
* Removed easy_install installation step from documentation (pip is 
  sufficient)
* Added a simple example of usage to docs/usage.rst
* Reduced the default GA population and generation size (to allow applications 
  that use the different parameters to run quickly)
* Modified tests to account for the new default population, generation size
* Added docstrings to all methods

v0.2.0
~~~~~~

**2014-07-04**

* First upload to pypi.
* Added changes made to HISTORY (pypi upload, new version)

v0.1.0
~~~~~~

**2014-06-23**

* Start of ``pyeasyga`` development.

**2014-07-03**

* Implemented all of basic GA functionality
* Fix issue with odd-numbered population that causes an off-by-one error in the 
  population size
* Set default ga selection function to tournament_selection
* Created examples to show how to use the library
* Start versioning (better late than never); copied jeffknupp’s 
  update_version.sh from `sandman <https://github.com/jeffknupp/sandman/>`_
 
  **selected versioning standard:**  major.minor.micro (e.g. 2.1.5)
  
  - major => big changes that can break compatibility
  - minor => new features
  - micro => bug fixes



tests/test_pyeasyga.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_pyeasyga
----------------------------------

Tests for `pyeasyga` module.
"""

import unittest
import copy
from operator import attrgetter

from six import next

from pyeasyga import pyeasyga


class TestPyeasyga(unittest.TestCase):

    def setUp(self):
        """Test Problem: Assume you can only carry 3 items to sell in your bag.
        Pick a combination of items that maximises your profit. The profit for
        each item is listed.
        """
        self.seed_data = [('apple', 15), ('banana', 10), ('carrot', 12),
                          ('pear', 5), ('mango', 8)]

        member_1 = pyeasyga.Chromosome([0, 1, 0, 1, 1])
        member_2 = pyeasyga.Chromosome([1, 1, 0, 0, 1])
        member_3 = pyeasyga.Chromosome([1, 1, 1, 1, 0])
        member_4 = pyeasyga.Chromosome([0, 1, 1, 1, 1])

        self.population = []
        self.population.append(member_1)
        self.population.append(member_2)
        self.population.append(member_3)
        self.population.append(member_4)

        self.ga = pyeasyga.GeneticAlgorithm(self.seed_data)
        self.ga.population_size = 10
        self.ga.generations = 10

        self.ga.fitness_function = lambda member, data: sum(
            [profit for (selected, (fruit, profit)) in
             zip(member, data) if selected and
             member.count(1) == 3])

        self.ga.selection_function = self.ga.tournament_selection

    def test_genetic_algorithm_initialisation_1(self):
        ''' Test default initialisation '''
        ga_1 = pyeasyga.GeneticAlgorithm(self.seed_data)

        assert ga_1.population_size == 50
        assert ga_1.generations == 100
        assert ga_1.elitism is True
        assert ga_1.crossover_probability == 0.8
        assert ga_1.mutation_probability == 0.2

    def test_genetic_algorithm_initialisation_2(self):
        ''' Test initialisation with specific values '''
        ga_2 = pyeasyga.GeneticAlgorithm(self.seed_data,
                                         population_size=235,
                                         generations=57,
                                         crossover_probability=0.9,
                                         mutation_probability=0.05)
        ga_2.elitism = False
        ga_2.maximise_fitness = False
        fruit, profit = ga_2.seed_data[1]

        assert len(ga_2.seed_data) == 5
        assert (fruit, profit) == ('banana', 10)
        assert ga_2.population_size == 235
        assert ga_2.generations == 57
        assert ga_2.elitism is False
        assert ga_2.maximise_fitness is False
        assert ga_2.crossover_probability == 0.9
        assert ga_2.mutation_probability == 0.05

    def test_chromosome_initialisation_1(self):
        chromosome = pyeasyga.Chromosome(['a', 'b', 'c'])

        assert chromosome.genes == ['a', 'b', 'c']
        assert chromosome.fitness == 0
        assert str(chromosome) == "(0, ['a', 'b', 'c'])"

    def test_chromosome_initialisation_2(self):
        chromosome = pyeasyga.Chromosome(['d', 'e', 'f'])
        chromosome.fitness = 20

        assert chromosome.genes == ['d', 'e', 'f']
        assert chromosome.fitness == 20
        assert str(chromosome) == "(20, ['d', 'e', 'f'])"

    def test_fitness_function(self):
        func = self.ga.fitness_function
        data = self.ga.seed_data

        assert func([0, 1, 0, 1, 1], data) == 23
        assert func([1, 1, 0, 0, 1], data) == 33
        assert func([1, 1, 1, 1, 0], data) == 0
        assert func([0, 1, 1, 1, 1], data) == 0

    def test_crossover_function(self):
        parent_1 = [1, 1, 1, 1, 1]
        parent_2 = [0, 0, 0, 0, 0]
        crossover = lambda x, y: ((x[:2] + y[2:]), (y[:2] + x[2:]))
        self.ga.crossover_function = crossover
        child_1, child_2 = self.ga.crossover_function(parent_1, parent_2)

        assert child_1 == [1, 1, 0, 0, 0]
        assert child_2 == [0, 0, 1, 1, 1]

    def test_mutate_function(self):
        ''' Check that the individual is not the same after mutation
            Check that only one bit in the individual is changed '''
        individual = [0, 1, 1, 0, 1]
        self.ga.mutate_function(individual)
        res = [x == y for (x, y) in zip(individual, [0, 1, 1, 0, 1])]

        assert individual != [0, 1, 1, 0, 1]
        assert res.count(False) == 1

    def test_selection_function_1(self):
        ''' Test tournament selection '''
        self.ga.current_generation = self.population
        self.ga.calculate_population_fitness()
        self.ga.tournament_size = 4
        self.ga.selection_function = self.ga.tournament_selection
        individual = self.ga.selection_function(self.ga.current_generation)

        assert individual.genes == [1, 1, 0, 0, 1]
        assert individual.fitness == 33
        assert len(individual.genes) == 5
        assert individual in self.population

    def test_selection_function_2(self):
        ''' Test tournament selection with tournament size 0'''
        self.ga.current_generation = self.population
        self.ga.calculate_population_fitness()
        self.ga.tournament_size = 0
        self.ga.selection_function = self.ga.tournament_selection
        individual = self.ga.selection_function(self.ga.current_generation)

        assert len(individual.genes) == 5
        assert individual in self.population

    def test_selection_function_3(self):
        ''' Test random selection '''
        self.ga.selection_function = self.ga.random_selection
        individual = self.ga.selection_function(self.population)

        assert len(individual.genes) == 5
        assert individual in self.population

    def test_create_individual(self):
        data = self.ga.seed_data
        individual = self.ga.create_individual(data)

        assert len(individual) == 5
        assert all([ind in (0, 1) for ind in individual if ind in (0, 1)])

    def test_create_initial_population(self):
        pop_size = self.ga.population_size
        self.ga.create_initial_population()
        initial_pop = self.ga.current_generation

        assert len(initial_pop) == pop_size
        assert isinstance(initial_pop[0], type(pyeasyga.Chromosome([1])))
        assert sum([member.fitness for member in initial_pop]) == 0

    def test_calculate_population_fitness(self):
        self.ga.current_generation = self.population
        self.ga.calculate_population_fitness()
        current_gen = self.ga.current_generation

        assert sum([member.fitness for member in current_gen]) > 0

    def test_rank_population(self):
        self.ga.create_initial_population()
        self.ga.calculate_population_fitness()
        new_population = copy.deepcopy(self.ga.current_generation)
        new_population.sort(key=attrgetter('fitness'), reverse=True)
        self.ga.rank_population()
        current_gen = self.ga.current_generation

        assert current_gen[0].fitness == new_population[0].fitness
        assert current_gen[1].fitness == new_population[1].fitness

    def test_create_new_population(self):
        """ Write more functional test """
        pop_size = self.ga.population_size
        self.ga.create_initial_population()
        self.ga.calculate_population_fitness()
        self.ga.rank_population()
        self.ga.create_new_population()
        current_gen = self.ga.current_generation

        assert len(current_gen) == pop_size
        assert isinstance(current_gen[0], type(pyeasyga.Chromosome([1])))

    def test_create_first_generation(self):
        """ Write more functional test """
        pop_size = self.ga.population_size
        self.ga.create_first_generation()
        current_gen = self.ga.current_generation

        assert len(current_gen) == pop_size
        assert isinstance(current_gen[0], type(pyeasyga.Chromosome([1])))

    def test_create_next_generation(self):
        pop_size = self.ga.population_size
        self.ga.create_first_generation()
        self.ga.create_next_generation()
        current_gen = self.ga.current_generation

        assert len(current_gen) == pop_size
        assert isinstance(current_gen[0], type(pyeasyga.Chromosome([1])))

    def test_run(self):
        self.ga.run()
        current_gen = self.ga.current_generation
        last_generation = self.ga.last_generation()

        assert len(current_gen) == self.ga.population_size
        assert isinstance(current_gen[0], type(pyeasyga.Chromosome([1])))
        assert isinstance(next(last_generation), type((1, [1, 1, 1, 1, 1])))
        assert len(next(last_generation)) == 2
        assert len(next(last_generation)[1]) == 5

    def test_best_individual(self):
        self.ga.create_first_generation()
        best_fitness, best_genes = self.ga.best_individual()

        assert best_fitness == self.ga.current_generation[0].fitness
        assert best_genes == self.ga.current_generation[0].genes

    def test_last_generation(self):
        self.ga.create_first_generation()
        last_generation = self.ga.last_generation()

        assert isinstance(next(last_generation), type((1, [1, 1, 1, 1, 1])))
        assert len(next(last_generation)) == 2
        assert len(next(last_generation)[1]) == 5

    def tearDown(self):
        pass

if __name__ == '__main__':
    unittest.main()


examples/8_queens.py
import random
from pyeasyga import pyeasyga

# setup seed data
seed_data = [0, 1, 2, 3, 4, 5, 6, 7]

# initialise the GA
ga = pyeasyga.GeneticAlgorithm(seed_data,
                               population_size=200,
                               generations=100,
                               crossover_probability=0.8,
                               mutation_probability=0.2,
                               elitism=True,
                               maximise_fitness=False)


# define and set function to create a candidate solution representation
def create_individual(data):
    individual = data[:]
    random.shuffle(individual)
    return individual

ga.create_individual = create_individual


# define and set the GA's crossover operation
def crossover(parent_1, parent_2):
    crossover_index = random.randrange(1, len(parent_1))
    child_1a = parent_1[:crossover_index]
    child_1b = [i for i in parent_2 if i not in child_1a]
    child_1 = child_1a + child_1b

    child_2a = parent_2[crossover_index:]
    child_2b = [i for i in parent_1 if i not in child_2a]
    child_2 = child_2a + child_2b

    return child_1, child_2

ga.crossover_function = crossover


# define and set the GA's mutation operation
def mutate(individual):
    mutate_index1 = random.randrange(len(individual))
    mutate_index2 = random.randrange(len(individual))
    index_val1 = individual[mutate_index1]
    index_val2 = individual[mutate_index2]
    index_val1, index_val2 = index_val2, index_val1

ga.mutate_function = mutate


# define and set the GA's selection operation
def selection(population):
    return random.choice(population)

ga.selection_function = selection


# define a fitness function
def fitness(individual, data):
    collisions = 0
    for item in individual:
        item_index = individual.index(item)
        for elem in individual:
            elem_index = individual.index(elem)
            if item_index != elem_index:
                if item - (elem_index - item_index) == elem \
                        or (elem_index - item_index) + item == elem:
                    collisions += 1
    return collisions

ga.fitness_function = fitness       # set the GA's fitness function
ga.run()                            # run the GA


# function to print out chess board with queens placed in position
def print_board(board_representation):
    def print_x_in_row(row_length, x_position):
        print '',
        for _ in xrange(row_length):
            print '---',
        print '\n|',
        for i in xrange(row_length):
            if i == x_position:
                print '{} |'.format('X'),
            else:
                print '  |',
        print ''

    def print_board_bottom(row_length):
        print '',
        for _ in xrange(row_length):
            print '---',

    num_of_rows = len(board_representation)
    row_length = num_of_rows    # rows == columns in a chessboard

    for row in xrange(num_of_rows):
        print_x_in_row(row_length, board_representation[row])

    print_board_bottom(row_length)
    print '\n'

# print the GA's best solution (valid only if there are no collisions)
if ga.best_individual()[0] == 0:
    print ga.best_individual()
    print_board(ga.best_individual()[1])
else:
    print None


examples/one_dimensional_knapsack.py
from pyeasyga import pyeasyga

# setup data
data = [{'name': 'box1', 'value': 4, 'weight': 12},
        {'name': 'box2', 'value': 2, 'weight': 1},
        {'name': 'box3', 'value': 10, 'weight': 4},
        {'name': 'box4', 'value': 1, 'weight': 1},
        {'name': 'box5', 'value': 2, 'weight': 2}]

ga = pyeasyga.GeneticAlgorithm(data)        # initialise the GA with data


# define a fitness function
def fitness(individual, data):
    values, weights = 0, 0
    for selected, box in zip(individual, data):
        if selected:
            values += box.get('value')
            weights += box.get('weight')
    if weights > 15:
        values = 0
    return values

ga.fitness_function = fitness               # set the GA's fitness function
ga.run()                                    # run the GA
print ga.best_individual()                  # print the GA's best solution


examples/multi_dimensional_knapsack.py
from pyeasyga import pyeasyga

# setup data
data = [(821, 0.8, 118), (1144, 1, 322), (634, 0.7, 166), (701, 0.9, 195),
        (291, 0.9, 100), (1702, 0.8, 142), (1633, 0.7, 100), (1086, 0.6, 145),
        (124, 0.6, 100), (718, 0.9, 208), (976, 0.6, 100), (1438, 0.7, 312),
        (910, 1, 198), (148, 0.7, 171), (1636, 0.9, 117), (237, 0.6, 100),
        (771, 0.9, 329), (604, 0.6, 391), (1078, 0.6, 100), (640, 0.8, 120),
        (1510, 1, 188), (741, 0.6, 271), (1358, 0.9, 334), (1682, 0.7, 153),
        (993, 0.7, 130), (99, 0.7, 100), (1068, 0.8, 154), (1669, 1, 289)]

ga = pyeasyga.GeneticAlgorithm(data)        # initialise the GA with data
ga.population_size = 200                    # increase population size to 200


# define a fitness function
def fitness(individual, data):
    weight, volume, price = 0, 0, 0
    for (selected, item) in zip(individual, data):
        if selected:
            weight += item[0]
            volume += item[1]
            price += item[2]
    if weight > 12210 or volume > 12:
        price = 0
    return price

ga.fitness_function = fitness               # set the GA's fitness function
ga.run()                                    # run the GA
print ga.best_individual()                  # print the GA's best solution


pyeasyga/pyeasyga.py
# -*- coding: utf-8 -*-
"""
    pyeasyga module

"""

import random
import copy
from operator import attrgetter

from six.moves import range


class GeneticAlgorithm(object):
    """Genetic Algorithm class.

    This is the main class that controls the functionality of the Genetic
    Algorithm.

    A simple example of usage:

    >>> # Select only two items from the list and maximise profit
    >>> from pyeasyga.pyeasyga import GeneticAlgorithm
    >>> input_data = [('pear', 50), ('apple', 35), ('banana', 40)]
    >>> easyga = GeneticAlgorithm(input_data)
    >>> def fitness (member, data):
    >>>     return sum([profit for (selected, (fruit, profit)) in
    >>>                 zip(member, data) if selected and
    >>>                 member.count(1) == 2])
    >>> easyga.fitness_function = fitness
    >>> easyga.run()
    >>> print easyga.best_individual()

    """

    def __init__(self,
                 seed_data,
                 population_size=50,
                 generations=100,
                 crossover_probability=0.8,
                 mutation_probability=0.2,
                 elitism=True,
                 maximise_fitness=True):
        """Instantiate the Genetic Algorithm.

        :param seed_data: input data to the Genetic Algorithm
        :type seed_data: list of objects
        :param int population_size: size of population
        :param int generations: number of generations to evolve
        :param float crossover_probability: probability of crossover operation
        :param float mutation_probability: probability of mutation operation

        """

        self.seed_data = seed_data
        self.population_size = population_size
        self.generations = generations
        self.crossover_probability = crossover_probability
        self.mutation_probability = mutation_probability
        self.elitism = elitism
        self.maximise_fitness = maximise_fitness

        self.current_generation = []

        def create_individual(seed_data):
            """Create a candidate solution representation.

            e.g. for a bit array representation:

            >>> return [random.randint(0, 1) for _ in range(len(data))]

            :param seed_data: input data to the Genetic Algorithm
            :type seed_data: list of objects
            :returns: candidate solution representation as a list

            """
            return [random.randint(0, 1) for _ in range(len(seed_data))]

        def crossover(parent_1, parent_2):
            """Crossover (mate) two parents to produce two children.

            :param parent_1: candidate solution representation (list)
            :param parent_2: candidate solution representation (list)
            :returns: tuple containing two children

            """
            index = random.randrange(1, len(parent_1))
            child_1 = parent_1[:index] + parent_2[index:]
            child_2 = parent_2[:index] + parent_1[index:]
            return child_1, child_2

        def mutate(individual):
            """Reverse the bit of a random index in an individual."""
            mutate_index = random.randrange(len(individual))
            individual[mutate_index] = (0, 1)[individual[mutate_index] == 0]

        def random_selection(population):
            """Select and return a random member of the population."""
            return random.choice(population)

        def tournament_selection(population):
            """Select a random number of individuals from the population and
            return the fittest member of them all.
            """
            if self.tournament_size == 0:
                self.tournament_size = 2
            members = random.sample(population, self.tournament_size)
            members.sort(
                key=attrgetter('fitness'), reverse=self.maximise_fitness)
            return members[0]

        self.fitness_function = None
        self.tournament_selection = tournament_selection
        self.tournament_size = self.population_size / 10
        self.random_selection = random_selection
        self.create_individual = create_individual
        self.crossover_function = crossover
        self.mutate_function = mutate
        self.selection_function = self.tournament_selection

    def create_initial_population(self):
        """Create members of the first population randomly.
        """
        initial_population = []
        for _ in range(self.population_size):
            genes = self.create_individual(self.seed_data)
            individual = Chromosome(genes)
            initial_population.append(individual)
        self.current_generation = initial_population

    def calculate_population_fitness(self):
        """Calculate the fitness of every member of the given population using
        the supplied fitness_function.
        """
        for individual in self.current_generation:
            individual.fitness = self.fitness_function(
                individual.genes, self.seed_data)

    def rank_population(self):
        """Sort the population by fitness according to the order defined by
        maximise_fitness.
        """
        self.current_generation.sort(
            key=attrgetter('fitness'), reverse=self.maximise_fitness)

    def create_new_population(self):
        """Create a new population using the genetic operators (selection,
        crossover, and mutation) supplied.
        """
        new_population = []
        elite = copy.deepcopy(self.current_generation[0])
        selection = self.selection_function

        while len(new_population) < self.population_size:
            parent_1 = copy.deepcopy(selection(self.current_generation))
            parent_2 = copy.deepcopy(selection(self.current_generation))

            child_1, child_2 = parent_1, parent_2
            child_1.fitness, child_2.fitness = 0, 0

            can_crossover = random.random() < self.crossover_probability
            can_mutate = random.random() < self.mutation_probability

            if can_crossover:
                child_1.genes, child_2.genes = self.crossover_function(
                    parent_1.genes, parent_2.genes)

            if can_mutate:
                self.mutate_function(child_1.genes)
                self.mutate_function(child_2.genes)

            new_population.append(child_1)
            if len(new_population) < self.population_size:
                new_population.append(child_2)

        if self.elitism:
            new_population[0] = elite

        self.current_generation = new_population

    def create_first_generation(self):
        """Create the first population, calculate the population's fitness and
        rank the population by fitness according to the order specified.
        """
        self.create_initial_population()
        self.calculate_population_fitness()
        self.rank_population()

    def create_next_generation(self):
        """Create subsequent populations, calculate the population fitness and
        rank the population by fitness in the order specified.
        """
        self.create_new_population()
        self.calculate_population_fitness()
        self.rank_population()

    def run(self):
        """Run (solve) the Genetic Algorithm."""
        self.create_first_generation()

        for _ in range(1, self.generations):
            self.create_next_generation()

    def best_individual(self):
        """Return the individual with the best fitness in the current
        generation.
        """
        best = self.current_generation[0]
        return (best.fitness, best.genes)

    def last_generation(self):
        """Return members of the last generation as a generator function."""
        return ((member.fitness, member.genes) for member
                in self.current_generation)


class Chromosome(object):
    """ Chromosome class that encapsulates an individual's fitness and solution
    representation.
    """
    def __init__(self, genes):
        """Initialise the Chromosome."""
        self.genes = genes
        self.fitness = 0

    def __repr__(self):
        """Return initialised Chromosome representation in human readable form.
        """
        return repr((self.fitness, self.genes))
